Auto indent on `insert_at_line_start`

CptPotato opened the pull request:
First attempt at #4859. This makes `I` and `A` automatically indent, but **only** if the line is empty. Otherwise it behaves exactly as before. I thought about also handling indentation when the line only contains white space, but I left this out for now. Either way, this is a breaking change.

I roughly followed the implementation of `open_above`/`open_below`. Though, I'm not sure if the implementation is very good, all the text editing functions in the code base are new to me.

Here's the result of `insert_at_line_start` / `I` with this change (the only difference is on the empty line):

| before | master | this PR |
|---|---|---|
| ![image](https://user-images.githubusercontent.com/3957610/216818723-ad0aaaaa-2e30-4e54-9175-3696f2d9c996.png) | ![image](https://user-images.githubusercontent.com/3957610/216831840-375f8f7d-1950-4690-9fae-8418a5d20283.png) | ![image](https://user-images.githubusercontent.com/3957610/216818729-336b309b-3bab-42c5-8712-f200786dcbbe.png) |

I tried to handle all edge cases (i.e. cursor is at document start/end, multiple selections) and so far it seems to work. ~~I also plan to make `insert_at_line_end` behave similarly~~.

Edit: `insert_at_line_end` now works the same.

closes #4859

pascalkuthe commented:
I haven't looked too closely at the implementation yet but this does seem very useful to me and like a better default too. Thanks for working on this!

That being said we might want to keep the old behaviour as a seperate command so people that don't like this change can rebind.

pascalkuthe commented on helix-term/src/commands.rs:
```diff
@@ -2659,9 +2661,85 @@ fn last_picker(cx: &mut Context) {
+        let cursor_line = text.char_to_line(cursor);
+
+        let doc_end = contents.len_chars();
+        if range.from() == doc_end {
+            ranges.push(Range::point(doc_end + offs));
+            return (doc_end, doc_end, None);
+        }
```
```suggestion
```

I don't think this is correct, the EOF position is not always it's own line. If the file doesn't end with a newline terminator the EOF is part of a normal line and `I` should work in that case too.
This condition is not even required because all functions in helix handle both the one past the end char index and one past the end line just fine (And I think we should also auto indent in that case because everything else would be inconsistent)

CptPotato replied:
Yup, this is not required anymore now.

pascalkuthe commented on helix-term/src/commands.rs:
```diff
@@ -2659,9 +2661,85 @@ fn last_picker(cx: &mut Context) {
+            offs += text.chars().count();
+            (line_end_index, line_end_index, Some(text.into()))
+        } else {
+            // move cursor to the first non-whitespace character of the current line
+            let first_non_ws = line_content
+                .chars()
+                .enumerate()
+                .find_map(|(i, c)| if !c.is_whitespace() { Some(i) } else { None })
+                .unwrap_or_default();
+
+            let pos = offs + text.line_to_char(cursor_line) + first_non_ws;
+            ranges.push(Range::point(pos));
+            (pos, pos, None)
```
```suggestion
	     let mut pos = text.line_to_char(cursor_line);
             if let Some(whitespace_off) = find_first_non_whitespace_char(text.line(line)) {
             	   pos += whitespace_off
	          range.put_cursor(text, pos, cx.editor.mode == Mode::Select)
	     }
	     ranges.push(range)
            (pos, pos, None)
```

This branch does not behave exactly the same as the old code and also contains quite a bit of duplication. I ported the above from `goto_first_nonwhitespace` and made some adjustements to fit your usecase here (note that you must change `|range|` to `|mut range|` for this to work

CptPotato replied:
Applied this with some slight changes (using `map` instead of `if let` and taking `offs` into account). I noticed one source of confusion was that the `Change`'s from and to position refers to positions before the transactions were applied, while the `ranges` are selections using the positions of the new resulting text.

pascalkuthe replied:
good point forgot about that

pascalkuthe commented on helix-term/src/commands.rs:
```diff
@@ -2659,9 +2661,85 @@ fn last_picker(cx: &mut Context) {
+        let line_content = line_without_line_ending(&text, cursor_line);
+        if line_content.len_chars() == 0 {
+            // line is empty => auto indent
+            let (line_end_index, line_end_offset_width) = if cursor_line == 0 {
+                (0, 0)
+            } else {
```
this condition seems unnecessay. I guess you special cased this for performance reasons? I don't think thats a concern here. Especially with my other suggested changes the branch below is free anyway.

CptPotato replied:
This was a left-over check to prevent a panic. Though this isn't required anymore :+1:.

pascalkuthe commented on helix-term/src/commands.rs:
```diff
@@ -2659,9 +2661,85 @@ fn last_picker(cx: &mut Context) {
+                (0, 0)
+            } else {
+                (
+                    line_end_char_index(&doc.text().slice(..), cursor_line),
```
```suggestion
                    cursor_line_start
```
Avoid recalculating that here the line is always empty (you checked above), so the result of `line_end_char_index` is always the same as the line start

CptPotato replied:
Done.

pascalkuthe commented on helix-term/src/commands.rs:
```diff
@@ -2659,9 +2661,85 @@ fn last_picker(cx: &mut Context) {
+                cursor_line,
+            );
+
+            let indent_len = indent.len();
+            let mut text = String::with_capacity(indent_len);
+            text.push_str(&indent);
```
```suggestion
```

This is unnecessary `indent` is already a String, just delete this and use `indent` instead of `text below

CptPotato replied:
Another left-over, removed.

pascalkuthe commented on helix-term/src/commands.rs:
```diff
@@ -2659,9 +2661,85 @@ fn last_picker(cx: &mut Context) {
+    let mut offs = 0;
+
+    let mut transaction = Transaction::change_by_selection(contents, selection, |range| {
+        let cursor = range.from();
+        let cursor_line = text.char_to_line(cursor);
```
I would also recommend calculating `cursor_line_start` as you jump trough quite a few hoops to calculate that in a more complex way then really necessary later.

CptPotato replied:
Thanks for the hint, I'm using this version now.

pascalkuthe commented on helix-term/src/commands.rs:
```diff
@@ -2659,9 +2661,85 @@ fn last_picker(cx: &mut Context) {
+            } else {
+                (
+                    line_end_char_index(&doc.text().slice(..), cursor_line),
+                    doc.line_ending.len_chars(),
```
`doc.line_ending` is not always equivalent to the line_ending on a line. A document can have mixed line endings. You don't need this anyway

CptPotato replied:
Removed this.

pascalkuthe commented on helix-term/src/commands.rs:
```diff
@@ -2659,9 +2661,85 @@ fn last_picker(cx: &mut Context) {
+            text.push_str(&indent);
+
+            // calculate new selection ranges
+            let pos = offs + line_end_index + line_end_offset_width;
+            ranges.push(Range::point((pos + indent_len).saturating_sub(1)));
+
+            offs += text.chars().count();
```
```suggestion
            let pos = offs + cursor_line_start ;
            let indent_width = indent.chars().count();
            ranges.push(pos + indent_width);
            offs += indent_width;
```

you are using byte indicies here (`indent_len`). You should be using char incidences as you do below. Also you are not considering select mode here.

CptPotato replied:
Done.

pascalkuthe commented on helix-term/src/commands.rs:
```diff
@@ -2659,9 +2661,85 @@ fn last_picker(cx: &mut Context) {
+    transaction = transaction.with_selection(Selection::new(ranges, selection.primary_index()));
+    doc.apply(&transaction, view.id);
+
+    // goto_first_nonwhitespace(cx);
+    // enter_insert_mode(cx);
```
these comments can be removed.

CptPotato replied:
Done.

pascalkuthe reviewed (changes requested):
The idea is good I think, the implementation can be simplified (and is not quite correct, see review comments)

CptPotato commented:
Thanks for the detailed review and pointers @pascalkuthe. I expected that I was doing a lot of these things sub-optimally. I'll try to address these issues soon.

pascalkuthe commented on helix-term/src/commands.rs:
```diff
@@ -2659,9 +2661,68 @@ fn last_picker(cx: &mut Context) {
+            (line_end_index, line_end_index, Some(indent.into()))
+        } else {
+            // move cursor to the first non-whitespace character of the current line
+            let pos_before = text.line_to_char(cursor_line);
```
this is the same as `cursor_line_start` so no need to recompute

pascalkuthe commented on helix-term/src/commands.rs:
```diff
@@ -2659,9 +2661,68 @@ fn last_picker(cx: &mut Context) {
+        let cursor_line = range.cursor_line(text);
+        let cursor_line_start = text.line_to_char(cursor_line);
+
+        let line_content = line_without_line_ending(&text, cursor_line);
+        if line_content.len_chars() == 0 {
```
You are not using `line_content` anywhere except to check if its width it 0. This can be done slightly more efficiently by doing:
```suggestion
        if line_end_char_index(text, cursor_line) == cursor_line_start {
```

pascalkuthe reviewed (changes requested):
Two minor nits otherwise the implementation now LGTM.

CptPotato commented:
I made the last suggested changes, thanks again.

I also created a second commit that adds the same behavior to `insert_at_line_end` by moving the indent code to a separate function.

CptPotato commented:
Should I look into creating a test for the new behavior?

pascalkuthe commented:
> Should I look into creating a test for the new behavior?

I believe there is currently no tests for the existing behaviour so I wouldn't consider it a blocking issue but more tests are always good so if you want to add them go ahead

CptPotato commented:
I added a test and found one remaining issue (which I fixed).

The test is a little ugly but it's the best I could do :eyes:
It checks both `insert_at_line_start`/`end` on all lines with this snippet:

```rust
fn foo() {
    if let Some(_) = None {

    }
 
}

fn bar() {

}
```
<sup>(there's one line with existing whitespace)</sup>

Is this okay or should I change the test?

CptPotato commented:
Looks like some recent changes are causing conflicts.

Let me know when somebody wants to review this and I'll resolve them.

pascalkuthe reviewed (dismissed):
I have been testing this for a bit and did a pretty thorough review a while ago. Once the bugfix release (23.3.1) is done we can resolve the conflicts here (should be pretty easy) I think we can land this

CptPotato commented:
Sorry for the long delay. I rebased onto latest master, should be good to go now.

the-mikedavis commented on helix-term/src/commands.rs:
```diff
@@ -2756,23 +2756,81 @@ fn last_picker(cx: &mut Context) {
+// If the line is not empty, move the cursor to the specified fallback position.
+fn insert_with_indent(
+    cx: &mut Context,
+    cursor_fallback: impl Fn(usize, usize, &RopeSlice) -> usize,
```
Rather than a function that the callers pass in, I think it would be more idiomatic for the callers to pass an enum of `Start`/`Beginning` or `End` and then to have this function do a match:

```rust
let cursor_fallback = match insert_at {
    Start => find_first_non_whitespace_char(text.line(cursor_line)).map(|ws_offset| ws_offset + cursor_line_start).unwrap_or(cursor_line_start),
    End => line_end_char_index(text, cursor_line),
};
```

It would be a little less flexible if `insert_with_indent` were to be used in more places but I don't see other use-cases for that function.

Tudyx replied:
A potential use case is if we want to have `xc` insert at indent

CptPotato replied:
Agree that this is a bit simpler, I changed it to an enum.

the-mikedavis reviewed (commented):
I have a small style nit but otherwise I think this looks great (especially the integration tests!)

pascalkuthe reviewed (approved):
thanks for working on this! This is a great QOL improvement
